keps/sig-api-machinery/4346-informer-metrics/README.md

```diff
@@ -0,0 +1,935 @@
+know that this has succeeded?
+-->
+
+- Add metrics for informer 
+- Expose informer reflector/queue/eventHandler metrics
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- It does not introduce breaking changes for controllers which use informer.
+- It does not modify core Kubernetes components which use informer.
+- It does not list all informer metrics, which can add as needed
+
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+- Introduce the informer metrics struct `informerMetrics` contains queue/eventHandler metrics
+- Introduce the informer metrics provider interface `informerMetricsProvider`, implement in `k8s.io/component-base/metrics`
+- Add an environment to enable informer metrics
+- Revert the deleted `reflectorMetrics`
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+Client-go informer create a RingGrowing `pendingNotifications` for every eventHandler. This RingGrowing will grow, but never shrink. An informer has some eventHandlers, it is hard to distinguish which `pendingNotifications` linked to a lot of object. The `pendingNotifications` metrics  will help developers to distinguish which eventHandler is slow.
+
+#### Story 2
+Users want to know how many times the reflector performes list action.
+
+#### Story 3
+It is hard to known how many item in informer queue/store. Add metrics for queue/store, it will help developers to find the number of pending deltas.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+N/A
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The informer metrics is disabled by default. When enable informer metrics, the newly added metrics will increase CPU/MEM usage. 
+
+If the metrics result memory leak, users can disable the informer metrics.
+
+
+## Design Details
```

Comment:
looks like a previous attempt failed on stops not removing metrics, thus producing a leak: https://github.com/kubernetes/kubernetes/issues/73587#issuecomment-459335009.  This should be explicitly handled by this KEP.

Reply:
Added description about mitigating memory usage.